### PR TITLE
Require CAPTCHA when users attempt to join more than 3 sections in 24 hours

### PR DESCRIPTION
**Pull request by drew-beckmen**
We want to require users to fill out a CAPTCHA if they attempt to join 3 or more sections in a day. This PR contains the backend methods needed to implement this task as well as the end-to-end user validation for the `/join/:section_code` endpoint.

 ### Security

- [Jira - Epic](https://codedotorg.atlassian.net/secure/RapidBoard.jspa?rapidView=21&modal=detail&selectedIssue=LP-1378)

### Future work

This PR will be followed by two additional PRs:

- An independent reCAPTCHA React component that can be used across the site where necessary
- The end-to-end integration of the backend logic with the newly created React component within `JoinSection.jsx`

## Links

- [jira](https://codedotorg.atlassian.net/secure/RapidBoard.jspa?rapidView=21&modal=detail&selectedIssue=LP-1644)

## Testing story

Unit tests added to `user_test.rb` to test functionality of new methods added to `user.rb`. Unit tests also added to `followers_controller_test.rb` to test changes to `followers_controller.rb` and `student_user_new.html.haml`

## Screenshots

**Scenario 1:** user has not attempted to join 3 sections in the past 24 hours
Screen Shot 2020-11-25 at 8 34 20 PM

**Scenario 2:** user has already attempted to join 3 sections in the past 24 hours
Screen Shot 2020-11-25 at 8 32 01 PM

**Scenario 3** user has already attempted to join 3 sections in the past 24 hours and tried to join another section without completing the captcha
Screen Shot 2020-11-25 at 8 29 57 PM

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/controllers/followers_controller.rb [hunk 1]
```diff
@@ -25,14 +24,23 @@ def student_register
...
     end
 
-    Retryable.retryable on: [Mysql2::Error, ActiveRecord::RecordNotUnique], matching: /Duplicate entry/ do
-      if @user.save && @section&.add_student(@user)
-        sign_in(:user, @user)
-        redirect_to root_path, notice: I18n.t('follower.registered', section_name: @section.name)
-        return
+    if @user && @user.display_captcha? && !verify_recaptcha
+      flash[:alert] = I18n.t('follower.captcha_required')
+      # Without concatenating the section code, user is forced to type in code again
```
Comment (jamescodeorg) [inline dashboard/app/controllers/followers_controller.rb]: (very minor) It might be clearer to write this comment in an affirmative tone ("Concatenate the section code so the user doesn't have to...")

File: dashboard/app/controllers/followers_controller.rb [hunk 2]
```diff
@@ -25,14 +24,23 @@ def student_register
       return render 'student_user_new', formats: [:html]
     end
 
-    Retryable.retryable on: [Mysql2::Error, ActiveRecord::RecordNotUnique], matching: /Duplicate entry/ do
-      if @user.save && @section&.add_student(@user)
-        sign_in(:user, @user)
-        redirect_to root_path, notice: I18n.t('follower.registered', section_name: @section.name)
-        return
+    if @user && @user.display_captcha? && !verify_recaptcha
```
Comment (jamescodeorg) [inline dashboard/app/controllers/followers_controller.rb:27]: Mostly curious: It looks like all of the other error redirects in this controller are done in load_section.  Is there a reason you put the captcha logic here instead of load_section?  (I'm not sure which is better, just curious around your thinking.)

File: dashboard/app/models/user.rb [hunk 3]
```diff
@@ -124,8 +124,6 @@ class User < ActiveRecord::Base
 
   acts_as_paranoid # use deleted_at column instead of deleting rows
 
-  scope :ignore_deleted_at_index, -> {from 'users IGNORE INDEX(index_users_on_deleted_at)'}
```
Comment (jamescodeorg) [inline dashboard/app/models/user.rb]: Just curious, why this change?

File: dashboard/app/models/user.rb [hunk 4]
```diff
@@ -2261,6 +2259,40 @@ def providers
     end
   end
 
+  def num_failed_section_attempts
+    properties['section_attempts'] || 0
+  end
+
+  def reset_section_attempts?
```
Comment (jamescodeorg) [inline dashboard/app/models/user.rb:2273]: I think a header comment on this function describing the reset logic would be useful.

File: dashboard/app/models/user.rb [hunk 5]
```diff
@@ -2261,6 +2259,40 @@ def providers
...
+    num_failed_section_attempts == 0 || (DateTime.now - DateTime.parse(properties['section_attempts_last_reset'])).to_i > 0
```
Comment (jamescodeorg) [inline dashboard/app/models/user.rb]: Why do we want to return true if num_failed_section_attempts == 0?

File: dashboard/app/models/user.rb [hunk 6]
```diff
@@ -2261,6 +2259,40 @@ def providers
...
+  end
+
+  # If 24 hours has passed, reset the section attempts value to zero.
```
Comment (jamescodeorg) [inline dashboard/app/models/user.rb]: Update comment?  (Actually, it's probably a good idea to double-check the header comments for each of the functions below as well.  I think some of the logic was moved around but the header comments may not have been updated?)

File: dashboard/config/locales/en.yml [hunk 7]
```diff
@@ -170,6 +170,8 @@ en:
     invalid_password: "Password must be at least 6 characters"
     mismatch_password: "The two passwords you entered do not match"
     student_consent: "I confirm that I have my parent or legal guardian's permission to use the Code.org services."
+    agree_us_website: "I agree to using a website based in the United States."
```
Comment (jamescodeorg) [inline dashboard/config/locales/en.yml]: Some of these changes don't look related to your change.  Maybe there's some merge funniness?

File: dashboard/test/controllers/followers_controller_test.rb [hunk 8]
```diff
@@ -315,4 +338,15 @@ class FollowersControllerTest < ActionController::TestCase
 
     assert_redirected_to admin_directory_path
   end
+
+  # Place test at bottom to avoid reCAPTCHA issues with pre-existing tests
+  test 'student_user_new displays error when joining 3 or more sections in 24 hours without completing captcha' do
+    Recaptcha.configuration.skip_verify_env.delete("test")
```
Comment (jamescodeorg) [inline dashboard/test/controllers/followers_controller_test.rb]: Is there a way to do this without modifying external state (e.g. via mocking)?  If not, let's at least revert this change at the end of the test.  (In general, unit tests should not modify external state.)

File: dashboard/test/controllers/followers_controller_test.rb [hunk 9]
```diff
@@ -69,6 +69,29 @@ class FollowersControllerTest < ActionController::TestCase
     assert_template 'followers/student_user_new'
   end
 
+  test "student_user_new increments join section attempts when signed in" do
```
Comment (jamescodeorg) [inline dashboard/test/controllers/followers_controller_test.rb:72]: Nice tests!

File: dashboard/app/controllers/followers_controller.rb [hunk 10]
```diff
@@ -64,6 +72,10 @@ def load_section
     # Note that we treat the section as not being found if the section user
     # (i.e., the teacher) does not exist (possibly soft-deleted) or is not a teacher
     unless @section && @section.user&.teacher?
+      # Adjust user properties to track failed section join attempts
+      if current_user
```
Comment (jamescodeorg) [inline dashboard/app/controllers/followers_controller.rb:76]: Hmmm.  Is there a loophole here where a non-signed-in user can keep trying to join a section?

**Review by jamescodeorg — COMMENTED** [review]
Mostly cosmetic feedback, but there is one important-ish comment about whether there's a loophole in followers_controller.rb.  (Sorry for not noticing that earlier!)

File: dashboard/app/models/user.rb (hunk 3 above)
Comment (drew-beckmen) [reply dashboard/app/models/user.rb]: Not sure what happened here - seems like something happened with Git along the way.

File: dashboard/app/controllers/followers_controller.rb (hunk 10 above)
Comment (drew-beckmen) [reply dashboard/app/controllers/followers_controller.rb:76]: Yes, this has been sitting in the back of my mind for a bit. If you hit `/join/:section_code` while not signed in, you need to fill out a form to create an account after which you will automatically join the section indicates by the section_code parameter in the route. We can't track this behavior with the current implementation because a new user would always start with zero section attempts, thus not requiring captcha validation. In my opinion, the only way to fully solve this issue is to stick another captcha at the bottom of the form where the user creates an account. Not sure we have the ability to implement any fancy logic if we go that route since the user won't yet exist; thus we can't store anything server-side.

File: dashboard/test/controllers/followers_controller_test.rb (hunk 8 above)
Comment (drew-beckmen) [reply dashboard/test/controllers/followers_controller_test.rb]: Agreed - this felt like a hack that would come back to bite later on. Not sure how we would be able to mock it since the `verify_recaptcha` method call comes directly from the recaptcha gem we use. For now, I will just revert the change after running the test. Please let me know if you have other ideas!

File: dashboard/config/locales/en.yml (hunk 7 above)
Comment (drew-beckmen) [reply dashboard/config/locales/en.yml]: Yep - not exactly sure what happened. Apologies for not noticing this before opening the PR. I put my questionable git skills to use to resolve the issue so hopefully it is fixed the next time i push.

File: dashboard/app/models/user.rb (hunk 5 above)
Comment (drew-beckmen) [reply dashboard/app/models/user.rb]: My code in the method `num_failed_section_attempts` originally returned 0 if properties['section_attempts'] was not initialized. However, I've changed it to return nil in this case and adjusted `reset_section_attempts?` accordingly; hopefully that improves clarity.

File: dashboard/app/controllers/followers_controller.rb (hunk 2 above)
Comment (drew-beckmen) [reply dashboard/app/controllers/followers_controller.rb:27]: My thought process was that the `load_section` method is responsible for finding the section and validating that the user is allowed to join the section. If any of those checks fail, we increment the section-attempts hash and allow `load_section` to redirect as it did before. Checking for captcha verification seemed like a check that would come after loading the section... Not sure if that logic checks out. I'm open to suggestions!

File: dashboard/app/models/user.rb [hunk 11]
```diff
@@ -2261,6 +2261,42 @@ def providers
...
+
+  # Users must complete a captcha when joining a section if they have joined 3 or more in the past 24 hours
+  def display_captcha?
+    if reset_section_attempts?
+      reset_failed_section_attempts
+    end
+    num_failed_section_attempts >= 3
+  end
+
+  # Failed section attemps reset every day
```
Comment (jmkulwik) [inline dashboard/app/models/user.rb]: NIT: This header comment seems to describe an implementation detail. It may make more sense right before line 2292 to describe the if statement. 

Reasoning: header comments generally describe the overall purpose and/or functionality of a method. The way it's currently written makes me think the purpose of `increment_section_attempts` is to reset the failed section attempts count each day.

File: dashboard/app/controllers/followers_controller.rb (hunk 10 above)
Comment (jamescodeorg) [reply dashboard/app/controllers/followers_controller.rb:76]: I think what you see after joining depends on whether the accounts are student-owned or teacher-owned?  In my test class where I use word logins, I get the "select a student" page when joining from a logged out page.

I'm wondering if a quick and reasonable mitigation here might be to always show the captcha if the user is not logged in.  (Feel free to follow-up on this separately from this PR if you prefer.)

File: dashboard/test/controllers/followers_controller_test.rb [hunk 12]
```diff
@@ -69,6 +69,42 @@ class FollowersControllerTest < ActionController::TestCase
...
+      post :student_register, params: {section_code: 'INVALID'}
+      @new_student.reload
+    end
+    assert_equal true, @new_student.display_captcha?
+  end
+
+  # Enable reCAPTCHA gem in test env for just this unit test.
+  # Restores original configuration after testing the assertion
+  test 'student_user_new displays error when joining 3 or more sections in 24 hours without completing captcha' do
+    Recaptcha.configuration.skip_verify_env.delete("test")
```
Comment (jamescodeorg) [inline dashboard/test/controllers/followers_controller_test.rb:98]: I poked around a bit and found this pattern for stubbing calls to an external library/service.  Maybe you can try stubbing out the call to verify_recaptcha entirely?  (If that doesn't work, what you have here is ok too.)

**Review by jamescodeorg — APPROVED** [review]

File: dashboard/test/controllers/followers_controller_test.rb (hunk 12 above)
Comment (drew-beckmen) [reply dashboard/test/controllers/followers_controller_test.rb:98]: @jamescodeorg Did you forget to paste a link here?

Comment (jamescodeorg) [reply dashboard/test/controllers/followers_controller_test.rb:98]: Ooops, here's the link: https://github.com/code-dot-org/code-dot-org/blob/f970daeba23a8630f5d470fe18349753befeabc2/lib/test/cdo/aws/s3.rb#L10.  Maybe you can stub verify_recaptcha the way list_objects_v2 is stubbed in that test.
